pallets/drop3/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+		pub fn set_admin(origin: OriginFor<T>, new: T::AccountId) -> DispatchResultWithPostInfo {
+			T::SetAdminOrigin::ensure_origin(origin)?;
+			Self::deposit_event(Event::AdminChanged { old_admin: Self::admin() });
+			<Admin<T>>::put(new);
+			// Do not pay a fee
+			Ok(Pays::No.into())
+		}
+
+		/// Approve a RewardPool proposal, must be called from admin
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::approve_reward_pool())]
+		pub fn approve_reward_pool(
+			origin: OriginFor<T>,
+			id: T::PoolId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(Some(sender) == Self::admin(), Error::<T>::RequireAdmin);
+
+			RewardPools::<T>::mutate(id, |pool| {
+				if let Some(mut p) = pool.take() {
+					p.approved = true;
+					*pool = Some(p);
+					Self::deposit_event(Event::RewardPoolApproved { id });
+					Ok(().into())
+				} else {
+					Err(Error::<T>::NoSuchRewardPool.into())
+				}
+			})
+		}
+
+		/// Reject a RewardPool proposal, must be called from admin
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::reject_reward_pool())]
+		pub fn reject_reward_pool(
+			origin: OriginFor<T>,
+			id: T::PoolId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(Some(sender) == Self::admin(), Error::<T>::RequireAdmin);
+			let pool = RewardPools::<T>::get(id).ok_or(Error::<T>::NoSuchRewardPool)?;
+			// a reward pool can't be rejected if it was approved earlier
+			ensure!(!pool.approved, Error::<T>::RewardPoolAlreadyApproved);
+
+			// slash a portion from reserved balance
+			let to_slash = T::SlashPercent::get() * pool.total;
+			let (_, unslashed) =
+				<pallet_balances::Pallet<T> as ReservableCurrency<_>>::slash_reserved(
+					&pool.owner,
+					to_slash,
+				);
+			// theoretically unslashed should be always 0, but just to play it safe
+			// no error is thrown even if unslashed is non-zero.
+			let actual_slashed = to_slash - unslashed;
+			RewardPools::<T>::mutate(id, |pool| {
+				if let Some(mut p) = pool.take() {
+					p.remain = p.remain.saturating_sub(actual_slashed);
+					*pool = Some(p);
+				}
+			});
+
+			Self::deposit_event(Event::RewardPoolRejected { id });
+			Self::deposit_event(Event::BalanceSlashed { who: pool.owner, amount: actual_slashed });
+
+			Self::unreserve_and_close_reward_pool(id)
+		}
+
+		/// Start a reward pool, can be called by admin or reward pool owner
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::start_reward_pool())]
+		pub fn start_reward_pool(
+			origin: OriginFor<T>,
+			id: T::PoolId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let mut pool = RewardPools::<T>::get(id).ok_or(Error::<T>::NoSuchRewardPool)?;
+			ensure!(
+				Some(sender.clone()) == Self::admin() || sender == pool.owner,
+				Error::<T>::RequireAdminOrRewardPoolOwner
+			);
+			ensure!(pool.approved, Error::<T>::RewardPoolUnapproved);
+
+			pool.started = true;
+			RewardPools::<T>::insert(id, pool);
```

Comment:
> Why insert the pool again here? I assume this will overwrite the old one, but wouldn't it be more efficient to just modify the flag itself (only one bit)?

as the getter only get a copy, so `pool.started = true;` only modifies the copy.
You either have to override it by inserting, or using mutate.

Reply:
> Yes, using mutate would be a better solution in my opinion, as we just modify a flag and we don't have to copy the whole pool twice. How do you think?

It's actually a cleaner solution.
Adjusted using `try_mutate`